apps/ouraring/ouraring.star

```diff
@@ -0,0 +1,193 @@
+
+def activityView(readiness_scores, activity_scores, sleep_scores):
+    avg = cal_average(activity_scores)
+    points = []
+    for index, score in enumerate(activity_scores):
+        points.append((index, score - avg))
+
+    return render.Column(
+        children = [render.Row(children = [readinessIcon(readiness_scores[-1], True), activityIcon(activity_scores[-1], False), sleepIcon(sleep_scores[-1], True)], main_align = "space_evenly", expanded = True), render.Plot(
+            data = points,
+            width = 64,
+            height = 19,
+            color = "#0f0",
+            color_inverted = "#F33",
+            fill = True,
+        )],
+        main_align = "space_between",
+        expanded = True,
+    )
+
+def readinessView(readiness_scores, activity_scores, sleep_scores):
+    avg = cal_average(readiness_scores)
+    points = []
+    for index, score in enumerate(readiness_scores):
+        points.append((index, score - avg))
+
+    return render.Column(
+        children = [render.Row(children = [readinessIcon(readiness_scores[-1], False), activityIcon(activity_scores[-1], True), sleepIcon(sleep_scores[-1], True)], main_align = "space_evenly", expanded = True), render.Plot(
+            data = points,
+            width = 64,
+            height = 19,
+            color = "#0f0",
+            color_inverted = "#F33",
+            fill = True,
+        )],
+        main_align = "space_between",
+        expanded = True,
+    )
+
+def sleepView(readiness_scores, activity_scores, sleep_scores):
+    avg = cal_average(sleep_scores)
+    points = []
+    for index, score in enumerate(sleep_scores):
+        points.append((index, score - avg))
+
+    return render.Column(
+        children = [render.Row(children = [readinessIcon(readiness_scores[-1], True), activityIcon(activity_scores[-1], True), sleepIcon(sleep_scores[-1], False)], main_align = "space_evenly", expanded = True), render.Plot(
+            data = points,
+            width = 64,
+            height = 19,
+            color = "#0f0",
+            color_inverted = "#F33",
+            fill = True,
+        )],
+        main_align = "space_between",
+        expanded = True,
+    )
+
+def errorView(message):
+    return render.WrappedText(
+        content = message,
+        width = 64,
+        color = "#fff",
+    )
+
+def main(config):
+    if "apikey" not in config:
+        return render.Root(child = errorView("no API key"))
+
+    if "days" not in config:
+        return render.Root(child = errorView("no days"))
+
+    now = time.now()
+    from_date = (now - time.parse_duration(str(int(config["days"]) * 24) + "h")).format("2006-01-02")
+    to_date = now.format("2006-01-02")
+
+    sleep_data = None
+    sleep_dto = cache.get("oura_sleep_data")
```

Comment:
The cache is scoped to your app, but not to individual users, so this call will simply return the most recently written cache record for this key (which could be from a different user). Including the API key in the cache key should be sufficient to solve this.

Reply:
(The same applies to other locations in the code where the cache is used)

Reply:
> The cache is scoped to your app, but not to individual users, so this call will simply return the most recently written cache record for this key (which could be from a different user). Including the API key in the cache key should be sufficient to solve this.

Thanks! I'll add that in.